File: Examples/iOS/AudioKitDemo/AudioKitDemo/Synthesis/Instruments/FMSynthesizer.m
//
//  FMSynthesizer.m
//  AudioKitDemo
//
//  Created by Aurelius Prochazka on 2/15/15.
//  Copyright (c) 2015 Aurelius Prochazka. All rights reserved.
//

#import "FMSynthesizer.h"

@implementation FMSynthesizer

- (instancetype)init
{
    self = [super init];
    if (self) {
        AKADSREnvelope *envelope = [[AKADSREnvelope alloc] initWithAttackDuration:akp(0.1)
                                                                    decayDuration:akp(0.1)
                                                                     sustainLevel:akp(0.5)
                                                                  releaseDuration:akp(0.3)
                                                                            delay:akp(0)];
        
        FMSynthesizerNote *note = [[FMSynthesizerNote alloc] init];
        AKFMOscillator *oscillator = [AKFMOscillator oscillator];
        oscillator.baseFrequency        = note.frequency;
        oscillator.carrierMultiplier    = [note.color scaledBy:akp(2)];
        oscillator.modulatingMultiplier = [note.color scaledBy:akp(3)];
        oscillator.modulationIndex      = [note.color scaledBy:akp(10)];
        oscillator.amplitude            = [envelope scaledBy:akp(0.25)];
        [self setAudioOutput:oscillator];
}
    return self;
}
@end

// -----------------------------------------------------------------------------
#  pragma mark - FMSynthesizer Note
// -----------------------------------------------------------------------------


@implementation FMSynthesizerNote

- (instancetype)init
{
    self = [super init];
    if (self) {
        _frequency = [self createPropertyWithValue:440 minimum:100 maximum:20000];
        _color     = [self createPropertyWithValue:0.0 minimum:0   maximum:1];
        self.duration.value = 1.0;
    }
    return self;
}

- (instancetype)initWithFrequency:(float)frequency color:(float)color
{
    self = [self init];
    if (self) {
        _frequency.value = frequency;
        _color.value = color;
    }
    return self;
}

@end

File: Examples/OSX/Swift/AudioKitDemo/AudioKitDemo/FMSynthesizer.swift
//
//  FMSynthesizer.swift
//  AudioKitDemo
//
//  Created by Nicholas Arner on 3/1/15.
//  Copyright (c) 2015 AudioKit. All rights reserved.
//

class FMSynthesizer: AKInstrument{
    
    override init() {
        super.init()
        
        // Note Properties
        let note = FMSynthesizerNote()
        
        let envelope = AKADSREnvelope(
            attackDuration:  0.1.ak,
            decayDuration:   0.1.ak,
            sustainLevel:    0.5.ak,
            releaseDuration: 0.3.ak,
            delay: 0.ak
        )
        
        let oscillator = AKFMOscillator()
        oscillator.baseFrequency        = note.frequency
        oscillator.carrierMultiplier    = note.color.scaledBy(2.ak)
        oscillator.modulatingMultiplier = note.color.scaledBy(3.ak)
        oscillator.modulationIndex      = note.color.scaledBy(10.ak)
        oscillator.amplitude            = envelope.scaledBy(0.25.ak)

        setAudioOutput(oscillator)
        
    }
}



class FMSynthesizerNote: AKNote {
    
    // Note Properties
    var frequency = AKNoteProperty(value: 440, minimum: 100, maximum: 20000)
    var color = AKNoteProperty(value: 0, minimum: 0, maximum: 1)
    
    override init() {
        super.init()
        addProperty(frequency)
        addProperty(color)
        
        // Optionally set a default note duration
        self.duration.setValue(1.0)
    }
    
    convenience init(frequency:Float, color:Float){
        self.init()
        
        self.frequency.setValue(frequency)
        self.color.setValue(color)
    }
}